io/src/main/scala/fs2/io/file/pulls.scala

```diff
@@ -19,11 +19,11 @@ object pulls {
   def readAllFromFileHandle[F[_]](chunkSize: Int)(h: FileHandle[F]): Pull[F, Byte, Unit] =
     _readAllFromFileHandle0(chunkSize, 0)(h)
 
-  private def _readAllFromFileHandle0[F[_]](chunkSize: Int, offset: Long)(h: FileHandle[F]): Pull[F, Byte, Unit] = for {
-    res <- Pull.eval(h.read(chunkSize, offset))
-    next <- res.fold[Pull[F, Byte, Unit]](Pull.done)(o => Pull.output(o) *> _readAllFromFileHandle0(chunkSize, offset + o.size)(h))
-  } yield next
```

Comment:
This `yield next` was the cause of the leak -- each time through this pull, we'd push a `map(identity)` in to the `FreeC`, which would accumulate on the return. In essence, the `yield` results in the recursive call being out of tail position.

Reply:
Thanks for the explanation.  It took me a while to realise what was going on here and I would have really struggled to find this.